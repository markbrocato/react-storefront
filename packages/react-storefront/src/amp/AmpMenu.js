/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component, Fragment } from 'react'
import { inject } from 'mobx-react'
import withStyles from '@material-ui/core/styles/withStyles'
import withTheme from '@material-ui/core/styles/withTheme'
import { Helmet } from "react-helmet"
import Link from '../Link'
import Typography from '@material-ui/core/Typography'
import classnames from 'classnames'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'
import ExpandLessIcon from '@material-ui/icons/ExpandLess'

export const styles = theme => ({
  root: {
    position: 'relative',
    marginTop: theme.headerHeight,
    height: `calc(100vh - ${theme.headerHeight}px)`,
    borderTop: `1px solid ${theme.palette.divider}`, 
    backgroundColor: theme.palette.background.paper,
    boxShadow: '10px 2px 10px -5px rgba(0, 0, 0, 0.2)',
    paddingBottom: '64px',
    '& h3': {
      backgroundColor: theme.palette.background.paper,
    },
    '& section[expanded] > h3': {
      color: theme.palette.secondary.contrastText,
      backgroundColor: theme.palette.secondary.main,
      borderColor: theme.palette.secondary.main
    },
    '& a': {
      color: theme.typography.body1.color
    },
    '& ~ div[class*="amphtml-sidebar-mask"]': {
      background: 'rgba(0, 0, 0, 0.5)',
      opacity: 1
    }
  },
  item: {
    padding: '12px 16px',
    borderColor: theme.palette.divider,
    borderWidth: '0 0 1px 0',
    borderStyle: 'solid',
    display: 'block',
    textDecoration: 'none'
  },
  group: {
    textTransform: 'uppercase'
  },
  toggle: {
    position: 'absolute',
    right: '18px',
    top: '10px',
  },
  expand: {
    display: 'block',
    'section[expanded] > h3 > &': {
      display: 'none'
    }
  },
  collapse: {
    display: 'none',
    'section[expanded] > h3 > &': {
      display: 'block'
    }
  }
})

/**
 * An AMP-compatible menu based on amp-sidebar.
 * 
 * In addition to the CSS classes that can be overridden of menu subcomponents, you can also
 * assign specific classes to individual menu items by specifying a value for the `className`
 * field on any instance of `MenuItemModel`.
 */
@withTheme()
@withStyles(styles, { name: 'RSFAmpMenu' })
@inject(({ app }) => ({ menu: app.menu }))
export default class AmpMenu extends Component {
  
  static defaultProps = {
    id: 'moov_menu'
  }

  render() {
    const { id, menu, classes, className, drawerWidth, rootHeader, rootFooter, align } = this.props
    const root = menu.levels.length ? menu.levels[0] : null

    return (
      <Fragment>
        <Helmet key="helmet">
          <script async custom-element="amp-sidebar" src="https://cdn.ampproject.org/v0/amp-sidebar-0.1.js"></script>
          <script async custom-element="amp-accordion" src="https://cdn.ampproject.org/v0/amp-accordion-0.1.js"></script>
          <style amp-custom>{`
            #${id} {
              width: ${drawerWidth}px;
            }
          `}</style>
        </Helmet>
        <amp-sidebar key="sidebar" id={id} class={classnames(className, classes.root)} layout="nodisplay" side={align}>
          { rootHeader }
          { this.renderItem(root) }
          { rootFooter }
        </amp-sidebar>
      </Fragment>
    )
  }

  renderItem(item, key) {
    if (item == null) {
      return null
    } else if (item.items) {
      return this.renderGroup(item, key)
    } else {
      return this.renderLeaf(item, key)
    }
  }

  renderGroup(item, key) {
    let { classes, ExpandIcon, CollapseIcon, theme } = this.props

    ExpandIcon = ExpandIcon || theme.ExpandIcon || ExpandMoreIcon
    CollapseIcon = CollapseIcon || theme.CollapseIcon || ExpandLessIcon

    const section = (
      <section key={key}>
        { item.text && (
          <Typography
            variant="body1"
            component="h3"
            className={classnames(classes.item, classes.group, item.className)}
          >
            {item.text}
            <CollapseIcon className={classnames(classes.toggle, classes.collapse)}/>
            <ExpandIcon className={classnames(classes.toggle, classes.expand)}/>
          </Typography>
        )}
        <div>
          { item.items.map((item, i) => this.renderItem(item, i)) }
        </div>
      </section>
    )

    if (item.root) {
      return section
    } else {
      return <amp-accordion key={key} disable-session-states>{ section }</amp-accordion>
    }
  }

  renderLeaf(item, key) {
    return (
      <Link
        key={key}
        className={classnames(this.props.classes.item, item.className)}
        to={item.url}
      >{item.text}</Link>
    )
  }

}
