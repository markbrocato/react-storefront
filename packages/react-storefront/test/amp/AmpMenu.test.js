/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import AmpMenu from '../../src/amp/AmpMenu'
import { mount } from 'enzyme'
import TestProvider from '../TestProvider'
import AppModelBase from '../../src/model/AppModelBase'

describe('AmpMenu', () => {
  it('should render', () => {
    const menu = {
      levels: [{
        root: true,
        items: [{
          text: "Category 1",
          items: [{
            text: "Subcategory 1",
            url: "/s/1"
          }, {
            text: "Subcategory 2",
            url: "/s/2"
          }]
        }, {
          text: "Category 2",
          items: [{
            text: "Subcategory 1",
            url: "/s/1"
          }, {
            text: "Subcategory 2",
            url: "/s/2"
          }]
        }]
      }]
    }
  
    expect(
      mount(
        <TestProvider app={{ menu, amp: true }}>
          <AmpMenu/>
        </TestProvider>
      )
    ).toMatchSnapshot()
  })

  describe('MenuItem.className', () => {
    const app = AppModelBase.create({
      location: {
        pathname: '/',
        search: '',
        hostname: 'localhost'
      },
      menu: {
        levels: [{
          root: true,
          items: [{
            text: "Group 1",
            className: 'group-1',
            items: [
              { text: 'Item 1', url: '/item1', className: 'item-1', items: [
                { text: 'Child 1', url: '/item1/child1', className: 'child-1' }
              ]}
            ]
          }]
        }]
      }
    })

    const wrapper = mount(
      <TestProvider app={app}>
        <AmpMenu/>
      </TestProvider>
    )

    expect(wrapper.exists('.group-1')).toBe(true)
    expect(wrapper.exists('.child-1')).toBe(true)
    expect(wrapper.exists('.item-1')).toBe(true)
  })
})