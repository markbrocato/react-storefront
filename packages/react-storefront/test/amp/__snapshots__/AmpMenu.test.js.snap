// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmpMenu should render 1`] = `
<TestProvider app={{...}}>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <WithTheme(WithStyles(inject-AmpMenu))>
        <WithStyles(inject-AmpMenu) theme={{...}}>
          <inject-AmpMenu theme={{...}} classes={{...}}>
            <AmpMenu theme={{...}} classes={{...}} menu={{...}} id="moov_menu">
              <HelmetWrapper defer={true} encodeSpecialCharacters={true}>
                <SideEffect(NullComponent) defer={true} encodeSpecialCharacters={true} script={{...}} style={{...}}>
                  <NullComponent defer={true} encodeSpecialCharacters={true} script={{...}} style={{...}} />
                </SideEffect(NullComponent)>
              </HelmetWrapper>
              <amp-sidebar id="moov_menu" class="RSFAmpMenu-root-52" layout="nodisplay" side={[undefined]}>
                <section>
                  <div>
                    <amp-accordion disable-session-states={true}>
                      <section>
                        <WithStyles(Typography) variant="body1" component="h3" className="RSFAmpMenu-item-53 RSFAmpMenu-group-54">
                          <Typography variant="body1" component="h3" className="RSFAmpMenu-item-53 RSFAmpMenu-group-54" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
                            <h3 className="MuiTypography-root-58 MuiTypography-body1-67 RSFAmpMenu-item-53 RSFAmpMenu-group-54">
                              Category 1
                              <pure(ExpandLessIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57">
                                <ExpandLessIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57">
                                  <WithStyles(SvgIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57">
                                    <SvgIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                                      <svg className="MuiSvgIcon-root-93 RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                        <path d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z" />
                                        <path fill="none" d="M0 0h24v24H0z" />
                                      </svg>
                                    </SvgIcon>
                                  </WithStyles(SvgIcon)>
                                </ExpandLessIcon>
                              </pure(ExpandLessIcon)>
                              <pure(ExpandMoreIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56">
                                <ExpandMoreIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56">
                                  <WithStyles(SvgIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56">
                                    <SvgIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                                      <svg className="MuiSvgIcon-root-93 RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                        <path d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z" />
                                        <path fill="none" d="M0 0h24v24H0z" />
                                      </svg>
                                    </SvgIcon>
                                  </WithStyles(SvgIcon)>
                                </ExpandMoreIcon>
                              </pure(ExpandMoreIcon)>
                            </h3>
                          </Typography>
                        </WithStyles(Typography)>
                        <div>
                          <WithStyles(inject-Link) className="RSFAmpMenu-item-53" to="/s/1">
                            <inject-Link className="RSFAmpMenu-item-53" to="/s/1" classes={{...}}>
                              <Link className="RSFAmpMenu-item-53" to="/s/1" classes={{...}} history={{...}} router={{...}} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-102 RSFAmpMenu-item-53" style={[undefined]} onClick={[Function]} href="https://localhost/s/1">
                                  Subcategory 1
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                          <WithStyles(inject-Link) className="RSFAmpMenu-item-53" to="/s/2">
                            <inject-Link className="RSFAmpMenu-item-53" to="/s/2" classes={{...}}>
                              <Link className="RSFAmpMenu-item-53" to="/s/2" classes={{...}} history={{...}} router={{...}} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-102 RSFAmpMenu-item-53" style={[undefined]} onClick={[Function]} href="https://localhost/s/2">
                                  Subcategory 2
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                        </div>
                      </section>
                    </amp-accordion>
                    <amp-accordion disable-session-states={true}>
                      <section>
                        <WithStyles(Typography) variant="body1" component="h3" className="RSFAmpMenu-item-53 RSFAmpMenu-group-54">
                          <Typography variant="body1" component="h3" className="RSFAmpMenu-item-53 RSFAmpMenu-group-54" theme={{...}} classes={{...}} align="inherit" color="default" gutterBottom={false} headlineMapping={{...}} noWrap={false} paragraph={false}>
                            <h3 className="MuiTypography-root-58 MuiTypography-body1-67 RSFAmpMenu-item-53 RSFAmpMenu-group-54">
                              Category 2
                              <pure(ExpandLessIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57">
                                <ExpandLessIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57">
                                  <WithStyles(SvgIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57">
                                    <SvgIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                                      <svg className="MuiSvgIcon-root-93 RSFAmpMenu-toggle-55 RSFAmpMenu-collapse-57" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                        <path d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z" />
                                        <path fill="none" d="M0 0h24v24H0z" />
                                      </svg>
                                    </SvgIcon>
                                  </WithStyles(SvgIcon)>
                                </ExpandLessIcon>
                              </pure(ExpandLessIcon)>
                              <pure(ExpandMoreIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56">
                                <ExpandMoreIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56">
                                  <WithStyles(SvgIcon) className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56">
                                    <SvgIcon className="RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56" classes={{...}} color="inherit" component="svg" fontSize="default" viewBox="0 0 24 24">
                                      <svg className="MuiSvgIcon-root-93 RSFAmpMenu-toggle-55 RSFAmpMenu-expand-56" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true" role="presentation">
                                        <path d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z" />
                                        <path fill="none" d="M0 0h24v24H0z" />
                                      </svg>
                                    </SvgIcon>
                                  </WithStyles(SvgIcon)>
                                </ExpandMoreIcon>
                              </pure(ExpandMoreIcon)>
                            </h3>
                          </Typography>
                        </WithStyles(Typography)>
                        <div>
                          <WithStyles(inject-Link) className="RSFAmpMenu-item-53" to="/s/1">
                            <inject-Link className="RSFAmpMenu-item-53" to="/s/1" classes={{...}}>
                              <Link className="RSFAmpMenu-item-53" to="/s/1" classes={{...}} history={{...}} router={{...}} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-102 RSFAmpMenu-item-53" style={[undefined]} onClick={[Function]} href="https://localhost/s/1">
                                  Subcategory 1
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                          <WithStyles(inject-Link) className="RSFAmpMenu-item-53" to="/s/2">
                            <inject-Link className="RSFAmpMenu-item-53" to="/s/2" classes={{...}}>
                              <Link className="RSFAmpMenu-item-53" to="/s/2" classes={{...}} history={{...}} router={{...}} location={{...}} anchorProps={{...}}>
                                <a data-moov-link="on" className="RSFLink-root-102 RSFAmpMenu-item-53" style={[undefined]} onClick={[Function]} href="https://localhost/s/2">
                                  Subcategory 2
                                </a>
                              </Link>
                            </inject-Link>
                          </WithStyles(inject-Link)>
                        </div>
                      </section>
                    </amp-accordion>
                  </div>
                </section>
              </amp-sidebar>
            </AmpMenu>
          </inject-AmpMenu>
        </WithStyles(inject-AmpMenu)>
      </WithTheme(WithStyles(inject-AmpMenu))>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;
